security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/JwtOAuth2Service.scala

```diff
@@ -47,16 +46,15 @@ class JwtOAuth2Service[
       case Invalid(jwtError) => Future.failed(OAuth2CompoundException(one(jwtError)))
     }
 
-  override def introspectAccessToken(accessToken: String): Future[Option[Instant]] = {
+  override def introspectAccessToken(accessToken: String): Future[IntrospectedAccessTokenData] = {
     if (accessTokenIsJwt) {
-      Future(accessToken)
-        .flatMap(accessToken => introspectJwtToken[AccessTokenClaims](accessToken))
-        .flatMap(claims =>
-          if (verifyAccessTokenAudience(claims)) {
-            Future.successful(claims.expirationTime)
-          } else
-            Future.failed(OAuth2CompoundException(one(OAuth2AccessTokenRejection("Invalid audience claim"))))
-        )
+      introspectJwtToken[AccessTokenClaims](accessToken).map { claims =>
+        if (verifyAccessTokenAudience(claims)) {
```

Comment:
can we move it to a private method? I mean, instead of `verifyAccessTokenAudience` there could be a method that checks the configuration and creates introspected data or throws the exception. The exception message could have a richer description of what is wrong (the human-readable version of the condition). I wanted to give a note here only about the message in the exception, but the condition is "too far" from the message to be sure that in the future when someone alters it, the message and condition won't fall apart.